### ISSUE TITLE
Cache Eloquent relationships

### ISSUE THREAD

**Issue by AliN11** [laravel/ideas#1433]
Hi. I searched the web and I think there is no way to cache an eloquent relationship.
This would be useful for example when I have posts and want to cache only their comments.

```PHP
// App\Models\Post.php

public function comments()
{
    return Cache::remember('comments', 180, function(){
        return $this->belongsTo('App\Models\Comment');
    })
}
```

```PHP
// This produces: You cannot serialize or unserialize PDO instances
$post->comments;
```

**Comment by AliN11** [comment]
@mfn Thanks for reviewing. It's simple. I would like to cache comments for posts. For example we have posts and comments in our website. A post has 40 comments and no new comment has been posted for a while. So it is better to cache these comments instead of fetching them from DB on every page load.

What you mentioned, works only for one page request. I mean if I refresh the page, nothing caches.

**Comment by drbyte** [comment]
@AliN11 perhaps the [laravel-model-caching](https://github.com/GeneaLabs/laravel-model-caching) package by @mikebronner will suit your needs.

**Comment by AliN11** [comment]
@drbyte It's cool. But it doesn't cache a specific relationship.

**Comment by mfn** [comment]
I'm still not clear on your goal but I'm trying to summarize it: you want to cache (all/specific?) relations on a model across requests?

**Comment by fico7489** [comment]
I understand what is an @AliN11 goal. for example if you have 1 post and that post have 20 comments, with this code:

    $post = Post::with('comments')->find(1);

two DB queries are excecuted : 

    select * from posts ...
    //and
    select * from comments

for same code from above, he want only one query

    select * from posts ...

and get comments from cache...

**Comment by Patryk27** [comment]
> So think of that what happens if our page has 300K views per day and you call `$post->comments` on every request.

Probably nothing scary, since simple `SELECT` queries utilizing indexes are pretty quick and should be automatically cached by the database engine anyway`.

One way or another: instead of sending 300K queries to MySQL / MariaDB / Postgres, you're now sending 300K queries to Redis or the filesystem, so nothing really changed.

**Comment by AliN11** [comment]
@Patryk27
You are right but I don't mean that. It was just an example. Examples are always simple :D
Sometimes before performing cache, data are processed and mixed with some other data and queries.

**Comment by Patryk27** [comment]
Well then - show a more complex example that _could_ be solved inside the framework :-)

As for now, I think that it seems more like a domain-related thing that should be handled entirely by the application, not framework.

**Comment by AliN11** [comment]
@Patryk27 I know there are some tricky ways, however `$post->comments` is impossible to cache.

**Comment by mfn** [comment]
> however `$post->comments` is impossible to cache

Did you try with a custom getter or are they overridden in such a case?

**Comment by AliN11** [comment]
@mfn I tried. If we override the `__get` method which exists in `Illuminate\Database\Eloquent\Model`, We can't use any other relationships like this: `$post->likes` anymore.

If we use `__get()` like bellow:
```PHP
public function comments()
{
    return $this->belongsTo('App\Models\Comment');
}


public function __get($name)
{
    return $name;
}
```
And if we try `$post->comments`, the *getter* executes and `comments()` method is ignored. It means that all relationships are ignored by adding `__get()`

**Comment by mfn** [comment]
Sorry I meant `getCommentsAttribute`, did you try that?

**Comment by PooyaE** [comment]
why you simply don't cache the final result?

```
return $users = Cache::remember('users-x', 60, function() {
    return Post::with('comments')->find(1);
});
```
the costs are equal with the way you want...
except Quantity of cache files...

**Comment by luke83** [comment]
Don't know if still relevant, anyway I think this could help in such scenario:

https://stackoverflow.com/a/56263511

**Comment by al-one** [comment]
https://github.com/laravel/framework/pull/35411

https://github.com/al-one/eloquent-super-relations
